synapse/util/metrics.py

```diff
@@ -289,21 +290,26 @@ class DynamicCollectorRegistry(CollectorRegistry):
 
     def __init__(self) -> None:
         super().__init__()
-        self._pre_update_hooks: Dict[str, Callable[[], None]] = {}
+        self._pre_update_hooks: Dict[str, List[Callable[[], None]]] = {}
 
     def collect(self) -> Generator[Metric, None, None]:
         """
         Collects metrics, calling pre-update hooks first.
         """
 
-        for pre_update_hook in self._pre_update_hooks.values():
-            pre_update_hook()
+        for pre_update_hooks in self._pre_update_hooks.values():
+            for pre_update_hook in pre_update_hooks:
+                pre_update_hook()
 
         yield from super().collect()
 
-    def register_hook(self, metric_name: str, hook: Callable[[], None]) -> None:
+    def register_hook(self, server_name: str, hook: Callable[[], None]) -> None:
         """
         Registers a hook that is called before metric collection.
         """
 
-        self._pre_update_hooks[metric_name] = hook
+        server_hooks = self._pre_update_hooks.setdefault(server_name, [])
+        server_hooks.append(hook)
+
+    def clear(self, server_name: str) -> None:
```

Comment:
We can match https://github.com/element-hq/synapse/pull/18791
```suggestion
    def unregister_hooks_for_homeserver_instance_id(self, homeserver_instance_id: str) -> None:
```

Reply:
Merged and aligned with your approach in #18791